src/main/java/org/broadinstitute/hellbender/tools/exome/cnlohcaller/CNLOHCaller.java

```diff
@@ -0,0 +1,752 @@
+
+            final List<double[][][]> responsibilitiesForSegsAsList = responsibilitiesForSegs.collect();
+
+            // Update E_ln_phi_k (Step 3)
+            logger.info("Updating E_ln_phi_k (iteration " + i + " of " + numIterations +  ") ... ");
+            state.seteLnPhiK(calc_E_ln_phi_k(state.getEAlpha(), responsibilitiesByRho));
+
+            // Update pis (Step 4)
+            logger.info("Updating pis (iteration " + i + " of " + numIterations +  ") ... ");
+            final double[] pis = calcPis(responsibilitiesByAllele1, responsibilitiesByAllele2);
+            final List<Integer> keys = new ArrayList<>(state.getCnToPiMap().keySet());
+            for (int p = 0; p < pis.length; p++) {
+                state.getCnToPiMap().put(keys.get(p), pis[p]);
+            }
+
+            // Update E_alpha (Step 5)
+            logger.info("Updating E_alpha (iteration " + i + " of " + numIterations + ") ... ");
+            final int minCN = state.getCnToPiMap().keySet().stream().min(Integer::compare).get();
+            final int maxCN = state.getCnToPiMap().keySet().stream().max(Integer::compare).get();
+            state.setEAlpha(calcEAlpha(state.geteLnPhiK(), minCN, maxCN));
+
+            // Update rhos (Step 6)
+            logger.info("Updating rhos (iteration " + i + ") ... ");
+            state.setRhos(calcNewRhos(segments, responsibilitiesForSegsAsList, state.getCnToPiMap(), state.getLambda(),
+                    state.getRhos(), ctx));
+
+            // This only really needs to  be done once, but is necessary here due to scoping issues that will cause
+            //  compiler issues.
+            cnlohCalls = createCNLoHCalls(segments, state, responsibilitiesForSegsAsList);
+        }
+        logger.info("ACNV segments bias, variance (around copy neutral only): " + segmentMeanBiasInCR + ", " + segmentMeanVarianceInCR);
+        logger.info("Output file is not adjusted for the bias and variance.  No user action required.");
+        return cnlohCalls;
+    }
+
+    private List<CNLOHCall> createCNLoHCalls(final List<ACNVModeledSegment> segments, final CNLOHCallerModelState state, final List<double[][][]> responsibilitiesForSegs) {
+
+        final List<CNLOHCall> cnlohCalls = segments.stream().map(s -> new CNLOHCall(s)).collect(Collectors.toList());
+
+        final List<Pair<Integer, Integer>> maxMiNiPerSeg = IntStream.range(0, segments.size()).boxed()
+                .map(s -> max2dIndices(sumOnlyFirstDimension(responsibilitiesForSegs.get(s))))
+                .collect(Collectors.toList());
+
+        final List<double[]> rhosForBestMN = IntStream.range(0, segments.size()).boxed()
+                .map(s -> getFirstDimensionArray(responsibilitiesForSegs.get(s), maxMiNiPerSeg.get(s).getFirst(), maxMiNiPerSeg.get(s).getSecond()))
+                .collect(Collectors.toList());
+
+        final List<Integer> maxRhoi = rhosForBestMN.stream().map(da -> maxIndex(da)).collect(Collectors.toList());
```

Comment:
In fact, why not eliminate the intermediate with `final double[] calledRhoPerSeg = rhosForBestMN.stream().mapToDouble(da -> state.getRhos()[maxIndex(da)]).toArray()`?